src/common/url-base64-functions.ts

```diff
@@ -0,0 +1,194 @@
+            )
+        case 3:
+            return _fromCharCode(
+                ((0x0f & context.charCodeAt(0)) << 12) |
+                    ((0x3f & context.charCodeAt(1)) << 6) |
+                    (0x3f & context.charCodeAt(2)),
+            )
+        default:
+            return _fromCharCode(
+                ((0x1f & context.charCodeAt(0)) << 6) | (0x3f & context.charCodeAt(1)),
+            )
+    }
+}
+const fromUTF16toUTF8 = (b: string) =>
+    b.replace(
+        /[\xC0-\xDF][\x80-\xBF]|[\xE0-\xEF][\x80-\xBF]{2}|[\xF0-\xF7][\x80-\xBF]{3}/g,
+        encodeUTF8,
+    )
+
+const b64_regexp = /^(?:[A-Za-z\d+\/]{4})*?(?:[A-Za-z\d+\/]{2}(?:==)?|[A-Za-z\d+\/]{3}=?)?$/
+
+const b64_tab = ((a) => {
+    const tab = {}
+    a.forEach((c, i) => (tab[c] = i))
+    return tab
+})(b64_chars)
+
+const atobPolyfill = (asc: string) => {
+    asc = asc.replace(/\s+/g, '')
+    if (!b64_regexp.test(asc)) throw new TypeError('malformed base64.')
+    asc += '=='.slice(2 - (asc.length & 3))
+    let bin = '',
+        r1,
+        r2,
+        u24
+    for (let i = 0; i < asc.length; ) {
+        u24 =
+            (b64_tab[asc.charAt(i++)] << 18) |
+            (b64_tab[asc.charAt(i++)] << 12) |
+            ((r1 = b64_tab[asc.charAt(i++)]) << 6) |
+            (r2 = b64_tab[asc.charAt(i++)])
+        bin +=
+            r1 === 64
+                ? _fromCharCode((u24 >> 16) & 255)
+                : r2 === 64
+                ? _fromCharCode((u24 >> 16) & 255, (u24 >> 8) & 255)
+                : _fromCharCode((u24 >> 16) & 255, (u24 >> 8) & 255, u24 & 255)
+    }
+    return bin
+}
+
+const cleanB64 = (s: string) => s?.replace(/[^A-Za-z0-9\+\/]/g, '')
+
+const toU8A =
+    typeof Uint8Array.from === 'function'
+        ? Uint8Array.from.bind(Uint8Array)
+        : (it, fn: (any) => number = (x) => x) =>
+              new Uint8Array(Array.prototype.slice.call(it, 0).map(fn))
+
+// https://developer.mozilla.org/en-US/docs/Web/API/atob
+const _atob = is_atob
+    ? (asc: string) => _atob(cleanB64(asc))
+    : is_buffer
+    ? (asc: string) => Buffer.from(asc, 'base64').toString('binary')
+    : atobPolyfill
+
+const toUint8Array = is_buffer
+    ? (a: string) => toU8A(Buffer.from(a, 'base64'))
+    : (a: string) => toU8A(_atob(a), (c) => c?.charCodeAt(0))
+
+const toUTF8 = is_buffer
+    ? (a: string) => Buffer.from(a, 'base64').toString('utf8')
+    : is_text_decoder
+    ? (a: string) => is_text_decoder.decode(toUint8Array(a))
+    : (a: string) => fromUTF16toUTF8(_atob(a))
+const _unURI = (a: string) => cleanB64(a?.replace(/[-_]/g, (m0) => (m0 == '-' ? '+' : '/')))
+
+// takes Base 64 string and returns encrypted string
+const encode = (src: string, urlsafe = false) =>
+    urlsafe ? defendURI(encodeString(src)) : encodeString(src)
+
+// takes encrypted string and returns decoded Base 64 string
+const decode = (src: string) => toUTF8(_unURI(src))
+
+// checks that string is valid for decode
+const isValid = (src: any) => {
```

Comment:
```suggestion
/**
 * @description checks that string is valid for decode,if type of is not string, validation will not pass
 * @param src if type of is not string, validation will not pass, this function only validates string params
 * @returns {boolean} the validity of the input string based on the provided regex
 */
const isValid = (src: unknown) => {
    if (typeof src !== 'string') return false
    const s = src?.replace(/\s+/g, '')?.replace(/={0,2}$/, '')
    return !/[^\s0-9a-zA-Z\+/]/.test(s) || !/[^\s0-9a-zA-Z\-_]/.test(s)
}
```
by documentation and comments I meant `jsDoc`, if you wanna more about it please check [this cheatsheet](https://devhints.io/jsdoc).

the `unknown` suggestion description: 
if you don't know the type of the argument, please use `unknown` but I think we can use `string` here, please check if you can remove the `any` :pray: 
please update other functions as well if it's makes sense ( some of the are just easy to understand ) 
if you had any questions, please let me know :pray:

Reply:
Done @michio-rework please check is this lind of docks enough for us? 🙏